src/Workspaces/Core/Desktop/Workspace/Host/TemporaryStorage/TemporaryStorageServiceFactory.ReferenceCountedDisposable.cs

```diff
@@ -0,0 +1,240 @@
+            /// The target of this reference. This value is initialized to a non-<see langword="null"/> value in the
+            /// constructor, and set to <see langword="null"/> when the current reference is disposed.
+            /// </summary>
+            /// <remarks>
+            /// <para>This value is only cleared in order to support cases where one or more references is garbage
+            /// collected without having <see cref="Dispose"/> called.</para>
+            /// </remarks>
+            private T _instance;
+
+            /// <summary>
+            /// The boxed reference count, which is shared by all references with the same <see cref="Target"/> object.
+            /// </summary>
+            /// <remarks>
+            /// <para>This field serves as the synchronization object for the current type, since it is shared among all
+            /// counted reference to the same target object. Accesses to <see cref="StrongBox{T}.Value"/> should only
+            /// occur when this object is locked.</para>
+            ///
+            /// <para>PERF DEV NOTE: A concurrent (but complex) implementation of this type with identical semantics is
+            /// available in source control history. The use of exclusive locks was not causing any measurable
+            /// performance overhead even on 28-thread machines at the time this was written.</para>
+            /// </remarks>
+            private readonly StrongBox<int> _boxedReferenceCount;
+
+            /// <summary>
+            /// Initializes a new reference counting wrapper around an <see cref="IDisposable"/> object.
+            /// </summary>
+            /// <remarks>
+            /// <para>The reference count is initialized to 1.</para>
+            /// </remarks>
+            /// <param name="instance">The object owned by this wrapper.</param>
+            public ReferenceCountedDisposable(T instance)
+            {
+                _instance = instance ?? throw new ArgumentNullException(nameof(instance));
+                _boxedReferenceCount = new StrongBox<int>(1);
+            }
+
+            private ReferenceCountedDisposable(T instance, StrongBox<int> referenceCount)
+            {
+                _instance = instance;
+
+                // The reference count has already been incremented for this instance
+                _boxedReferenceCount = referenceCount;
+            }
+
+            /// <summary>
+            /// Gets the target object.
+            /// </summary>
+            /// <remarks>
+            /// <para>This call is not valid after <see cref="Dispose"/> is called. If this property or the target
+            /// object is used concurrently with a call to <see cref="Dispose"/>, it is possible for the code to be
+            /// using a disposed object. After the current instance is disposed, this property returns
+            /// <see langword="null"/>. However, the exact time when this property starts returning null after
```

Comment:
@sharwell Well, but that's the caller's mishandling of concurrency, not yours. The caller could also just write (in a single threaded way):

    var blah = reference.Target;
    reference.Dispose();
    // use blah

which from the perspective of your type is the same race you're worried about. Put another way, that's the *caller* being unsynchronized in their calls, which are invalid regardless of the guarantees you make for your type.

I suspect the check works out to be two x86 instructions that might add up to nearly no measurable time, especially given the uses of this type are doing things like creating memory mapped files and messing with virtual memory. The cost to us of finding a bug will be orders and orders of magnitude larger. I'd just add the check.